src/Orchard.Environment.Extensions/Compilers/CSharpExtensionCompiler.cs

```diff
@@ -75,46 +138,147 @@ public bool Compile(ProjectContext context, string config, string buildBasePath)
+                                path = ResolveAssetPath(outputPath, probingFolderPath, fileName);
+                            }
+                        }
+
+                        if (!String.IsNullOrEmpty(path))
+                        {
+                            references.Add(path);
+                        }
+                    }
+                }
+                // Check for a precompiled library
+                else if (library != null && !dependency.CompilationAssemblies.Any())
+                {
+                    var projectContext = ProjectContext.CreateContextForEachFramework(library.Project.ProjectDirectory).FirstOrDefault();
+
+                    if (projectContext != null)
+                    {
+                        var fileName = library.Identity.Name + FileNameSuffixes.DotNet.DynamicLib;
+
+                        // Search in the precompiled project output path
+                        var path = Path.Combine(projectContext.GetOutputPaths(config).CompilationOutputPath, fileName);
+
+                        if (!File.Exists(path))
+                        {
+                            // Fallback to this project output path or probing folder
+                            path = ResolveAssetPath(outputPath, probingFolderPath, fileName);
+                        }
+
+                        if (!String.IsNullOrEmpty(path))
+                        {
+                            references.Add(path);
+                        }
+                    }
+                }
+                // Check for an ambient library
+                else if (library != null && _ambientLibraries.TryGetValue(library.Identity.Name, out compilationResult))
+                {
+                    references.AddRange(dependency.CompilationAssemblies.Select(r => Path.Combine(runtimeDirectory, r.FileName)));
+                }
+                else
+                {
+                    references.AddRange(dependency.CompilationAssemblies.Select(r => r.ResolvedPath));
+                }
             }
 
+            // Check again if already compiled, here through the dependency graph
+            if (_compiledLibraries.TryGetValue(context.RootProject.Identity.Name, out compilationResult))
+            {
+                return _compiledLibraries[context.RootProject.Identity.Name];
+            }
+
+            var sw = Stopwatch.StartNew();
+
             var resources = new List<string>();
+
+            string depsJsonFile = null;
+            DependencyContext dependencyContext = null;
+
+            // Add dependency context as a resource
             if (compilationOptions.PreserveCompilationContext == true)
             {
-                var allExports = exporter.GetAllExports().ToList();
-                var dependencyContext = new DependencyContextBuilder().Build(compilationOptions,
+                var allExports = exporter.GetAllExports().Where(x => x.Library.Compatible).ToList();
```

Comment:
Yes, you're right, i'll do it. Maybe other places to do the same ... I planed to do this kind of things but because i'm not sure all of this will be used ...